File: backend/app/models/recurring_template.py
"""
Recurring Template Model

Modèle pour les transactions récurrentes.
Génère automatiquement des transactions projetées selon la fréquence.
"""
from sqlalchemy import Column, String, Numeric, Date, Integer, Enum, ForeignKey, DateTime, Text, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from datetime import date
import enum
import uuid

from app.database import Base


class Frequency(str, enum.Enum):
    """Fréquence de récurrence"""
    WEEKLY = "WEEKLY"  # Hebdomadaire
    MONTHLY = "MONTHLY"  # Mensuelle
    QUARTERLY = "QUARTERLY"  # Trimestrielle (tous les 3 mois)
    YEARLY = "YEARLY"  # Annuelle
    CUSTOM = "CUSTOM"  # Personnalisée (utilise custom_days)


class RecurringTemplate(Base):
    """
    Template de transaction récurrente
    
    Génère des transactions projetées automatiquement selon :
    - Fréquence (WEEKLY, MONTHLY, QUARTERLY, YEARLY, CUSTOM)
    - Date de début (start_date)
    - Date de fin optionnelle (end_date)
    - Jour du mois pour MONTHLY/QUARTERLY/YEARLY (day_of_month)
    - Jour de la semaine pour WEEKLY (day_of_week: 0=Lundi, 6=Dimanche)
    - Nombre de jours pour CUSTOM (custom_days)
    """
    __tablename__ = "recurring_templates"

    # Identifiant
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))

    # Informations de base (similaires à Transaction)
    name = Column(String(255), nullable=False)
    amount = Column(Numeric(12, 2), nullable=False)
    type = Column(String(50), nullable=False)  # INCOME, EXPENSE, TRANSFER
    description = Column(Text, nullable=True)

    # Configuration de récurrence
    frequency = Column(Enum(Frequency), nullable=False)
    start_date = Column(Date, nullable=False)  # Date de première occurrence
    end_date = Column(Date, nullable=True)  # Date de fin (None = indéfini)

    # Paramètres spécifiques selon fréquence
    day_of_month = Column(Integer, nullable=True)  # 1-31 pour MONTHLY/QUARTERLY/YEARLY
    day_of_week = Column(Integer, nullable=True)  # 0-6 pour WEEKLY (0=Lundi)
    custom_days = Column(Integer, nullable=True)  # Nombre de jours pour CUSTOM

    # Relations
    household_id = Column(String, ForeignKey("households.id", ondelete="CASCADE"), nullable=False)
    account_id = Column(String, ForeignKey("accounts.id", ondelete="RESTRICT"), nullable=False)
    destination_account_id = Column(String, ForeignKey("accounts.id", ondelete="RESTRICT"), nullable=True)
    category_id = Column(String, ForeignKey("categories.id", ondelete="SET NULL"), nullable=True)

    # État
    is_active = Column(String, nullable=False, default="true")  # "true"/"false"
    
    # Audit
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now(), nullable=True)

    # Relationships
    household = relationship("Household", back_populates="recurring_templates")
    account = relationship("Account", foreign_keys=[account_id])
    destination_account = relationship("Account", foreign_keys=[destination_account_id])
    category = relationship("Category")

    def __repr__(self):
        return f"<RecurringTemplate(id={self.id}, name={self.name}, frequency={self.frequency}, amount={self.amount})>"

File: frontend/src/types/recurringTemplate.ts
/**
 * Recurring Template Types
 * 
 * Types pour les templates de transactions récurrentes.
 */

export enum Frequency {
  WEEKLY = "WEEKLY",
  MONTHLY = "MONTHLY",
  QUARTERLY = "QUARTERLY",
  YEARLY = "YEARLY",
  CUSTOM = "CUSTOM"
}

export const FrequencyLabels: Record<Frequency, string> = {
  [Frequency.WEEKLY]: "Hebdomadaire",
  [Frequency.MONTHLY]: "Mensuelle",
  [Frequency.QUARTERLY]: "Trimestrielle",
  [Frequency.YEARLY]: "Annuelle",
  [Frequency.CUSTOM]: "Personnalisée"
};

export enum TransactionType {
  INCOME = "INCOME",
  EXPENSE = "EXPENSE",
  TRANSFER = "TRANSFER"
}

export interface RecurringTemplate {
  id: string;
  name: string;
  amount: number;
  type: TransactionType;
  description: string | null;
  
  frequency: Frequency;
  start_date: string; // ISO date
  end_date: string | null; // ISO date
  
  day_of_month: number | null; // 1-31 for MONTHLY/QUARTERLY/YEARLY
  day_of_week: number | null; // 0-6 for WEEKLY (0=Lundi)
  custom_days: number | null; // Number of days for CUSTOM
  
  account_id: string;
  destination_account_id: string | null;
  category_id: string | null;
  household_id: string;
  
  is_active: string; // "true" | "false"
  created_at: string;
  updated_at: string | null;
}

export interface RecurringTemplateCreate {
  name: string;
  amount: number;
  type: TransactionType;
  description?: string;
  
  frequency: Frequency;
  start_date: string; // YYYY-MM-DD
  end_date?: string | null; // YYYY-MM-DD
  
  day_of_month?: number | null;
  day_of_week?: number | null;
  custom_days?: number | null;
  
  account_id: string;
  destination_account_id?: string | null;
  category_id?: string | null;
}

export interface RecurringTemplateUpdate {
  name?: string;
  amount?: number;
  description?: string;
  
  end_date?: string | null;
  
  day_of_month?: number | null;
  day_of_week?: number | null;
  custom_days?: number | null;
  
  category_id?: string | null;
  is_active?: string; // "true" | "false"
}

export interface BulkCancelRequest {
  start_date: string; // YYYY-MM-DD
  end_date: string; // YYYY-MM-DD
}

export interface BulkUpdateRequest {
  start_date: string; // YYYY-MM-DD
  end_date: string; // YYYY-MM-DD
  amount: number;
}

// Helper function pour obtenir le label d'une fréquence
export function getFrequencyLabel(frequency: Frequency): string {
  return FrequencyLabels[frequency];
}

// Helper function pour obtenir les jours de la semaine
export const WeekDays = [
  { value: 0, label: "Lundi" },
  { value: 1, label: "Mardi" },
  { value: 2, label: "Mercredi" },
  { value: 3, label: "Jeudi" },
  { value: 4, label: "Vendredi" },
  { value: 5, label: "Samedi" },
  { value: 6, label: "Dimanche" }
];

// Helper function pour formatter la récurrence en texte
export function formatRecurrence(template: RecurringTemplate): string {
  const freq = FrequencyLabels[template.frequency];
  
  if (template.frequency === Frequency.WEEKLY && template.day_of_week !== null) {
    const day = WeekDays[template.day_of_week].label;
    return `${freq} le ${day}`;
  }
  
  if (template.frequency === Frequency.MONTHLY && template.day_of_month !== null) {
    return `${freq} le ${template.day_of_month}`;
  }
  
  if (template.frequency === Frequency.QUARTERLY && template.day_of_month !== null) {
    return `${freq} le ${template.day_of_month}`;
  }
  
  if (template.frequency === Frequency.YEARLY && template.day_of_month !== null) {
    const startDate = new Date(template.start_date);
    const month = startDate.toLocaleDateString('fr-FR', { month: 'long' });
    return `${freq} le ${template.day_of_month} ${month}`;
  }
  
  if (template.frequency === Frequency.CUSTOM && template.custom_days !== null) {
    return `Tous les ${template.custom_days} jours`;
  }
  
  return freq;
}
